### ISSUE TITLE
Values logged in test_epoch_end not returned when calling test()

### ISSUE THREAD

**Issue by pbmstrk** [PyTorchLightning/pytorch-lightning#4234]
## 🐛 Bug

When calling `test()`, if values are logged only in the `test_epoch_end` method they are not returned. This leads to the following somewhat inconsistent behaviour:
- Values logged only in `step` method -> appear in list returned by `test()`.
- Values logged in `step` and `epoch_end` -> both appear in list returned by `test()`.
- Values logged only in `epoch_end` -> values do not appear.

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1C1UiT815EAP6CGZcumhQ_JP8RMVBdFpX?usp=sharing

### Expected behavior

If values are logged only in `epoch_end` they should be returned by `test()`.

### Environment

PL version: 1.0.2

see colab.

**Comment by Borda** [comment]
@SeanNaren mind have look? :]

**Comment by SeanNaren** [comment]
Thanks for the report @pbmstrk! Fairly simple but with a couple of ways to fix this. 

The issue arises because when we log metrics, we do not include any custom logged epoch end metrics. The reason why this is included when something happens within the `test_step` function is because we actually complete the loop [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector.py#L174) which reduces step metrics per dataloader.

At [this](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector.py#L197) line we add everything in callback_metrics which includes our custom logged metric in `test_epoch_end`.

**Solution 1**

Add the below lines in the step metrics calculation in `logger_connector.py`, replacing the simple `continue` operation:

```python
if len(self.callback_metrics) > 0:
    self.eval_loop_results.append(self.callback_metrics)
continue
```
The issue with this solution is that if we had multiple data-loaders, our custom metric will be added to every result for each data-loader.

will return with multi dl:

```python
[
    {
        'y/dl1': 1.4e-4,
        'custom_metric': 0.4
    },
    {
        'x/dl2': 1.8e-3,
        'custom_metric': 0.4
    }
]
```
**Solution 2**

Add the below lines after the step metrics calculation, creating a separate entry in the return Results List for custom metrics. I.e add after the dataloader metrics loop:
```python
                if len(self.callback_metrics) > 0:
                    self.eval_loop_results.append(self.callback_metrics)
```

will return:

```python
[
    {
        'y/dl1': 1.4e-4,
    },
    {
        'x/dl2': 1.8e-3,
    },
    {
        'custom_metric': 0.4
    }
]

```

**Comment by SeanNaren** [comment]
Will implement solution 1, as this reflects what we see if there are step logging.

**Comment by Enolerobotti** [comment]
Hi @SeanNaren! 
It seems this issue happens again starting from PL version>=1.5.5. Could you please have a look?
